## *Math For Machine Learning* 
### Chapter 3 (Analytic Geometry): Exercises

In [1]:
import numpy as np

#### Exercise 3.5: 
We have a Euclidean vector space \\(\mathbb{R}^5\\) with the dot product. Given subspace \\(U\subseteq\mathbb{R}^5\\) and \\(x\in\mathbb{R}^5\\) as defined below:

In [10]:
U = np.array(
    [
        [0, 1, -3, -1],
        [-1, -3, 4, -3],
        [2, 1, 1, 5],
        [0, -1, 2, 0], 
        [2, 2, 1, 7]
    ]
)

x = np.array([-1, -0, -1, 4, 1])

a. *Determine the orthogonal projection \\(\pi_U(x)\\) of \\(x\\) onto \\(U\\).* \
Step 1: since we do not know whether the columns are linearly-independent, much less orthonormal, we can compute the pseudoinverse of \\(U\\).

In [11]:
Up = np.linalg.pinv(U)

Step 2: use the pseudoinverse \\(U^+\\) to form the projection: \\(\pi_U(x) = UU^+x\\).

In [15]:
pi_x = U @ Up @ x.T
print(f"The projection pi(x) onto U is: {pi_x}")

The projection pi(x) onto U is: [-1.57142857  1.14285714 -0.42857143  0.57142857  1.        ]



b. *Determine the distance \\(d(x, U)\\).* \
Step 1: because \\(\pi_U(x)\\) is the closest point in \\(U\\) to \\(x\\), we first need the vector \\(v\\) from \\(\pi_U(x)\to x\\) with \\(v = x - \pi_U(x)\\).

In [16]:
v = pi_x - x

Step 2: now we just need \\(d = \lVert v\rVert\\).

In [20]:
d = np.linalg.norm(v)
print(f"The value of d(x, U): {d}")

The value of d(x, U): 3.703280399090206


#### Exercise 3.6:
Consider \\(\mathbb{R}^3\\) with the inner product: \\(\langle x, y\rangle := x^TBy\\). \\(B\\) follows:

In [24]:
B = np.array(
    [
        [2, 1, 0],
        [1, 2, -1],
        [0, -1, 2]
    ]
)

\\(e_1, e_2, e_3\\) are the standard, canonical basis for \\(\mathbb{R}^3\\). 

a. *Determine the orthogonal projection \\(\pi_U(e_2)\\) of \\(e_2\\) onto \\(U=\text{span}[e_1, e_3]\\).*

In [25]:
U = np.array(
    [
        [1, 0],
        [0, 0],
        [0, 1]
    ]
)

We can see immediately that \\(B\\) is symmetric. Now we check for whether it is positive definite by verifying that its eigenvalues are all positive: 

In [26]:
eigenvals, _ = np.linalg.eig(B)
all_pos = all(val > 0 for val in eigenvals)
print(f"all eigenvalues are positive: {all_pos}")

all eigenvalues are positive: True


We also know that the columns of \\(U\\) are linearly independent. This means that \\(U^TBU\\) is invertible, giving us the projection matrix: \\(P_B = U(U^TBU)^{-1}U^TB\\), which will allow us to find the project point \\(p_{e_2} = P_Be_2\\).

In [31]:
P_B = U @ np.linalg.inv(U.T @ B @ U) @ U.T @ B
e2 = np.array([0, 1, 0])
pe2 = P_B @ e2
print(f"The projection of e2 onto U is: {pe2}")

The projection of e2 onto U is: [ 0.5  0.  -0.5]


b. *Compute \\(d(e_2, U)\\)*. \
We need \\(\lVert e_2 - p_{e_2} \rVert_B\\), the \\(B\\)-norm: for \\(r = e_2 - p_{e_2}\\), it's: \\(\sqrt{r^TBr}\\).

In [34]:
r_e2 = e2 - pe2
print(f"The residual vector is: {r_e2}")
d_e2 = np.sqrt(r_e2.T @ B @ r_e2)
print(f"The distance from e2 to U is: {d_e2}")

The residual vector is: [-0.5  1.   0.5]
The distance from e2 to U is: 1.0


#### Exercise 3.7:

Let \\(V\\) be a vector space and \\(\pi\\) an endomorphism (\\(V\to V\\) mapping).

a. *Prove that \\(\pi\\) is a projection iff \\(\text{id}_V - \pi\\) is a projection, where \\(\text{id}_V\\) is the identity endomorphism on \\(V\\).* 

Approach and key insight: given that a projection is an idempotent mapping (i.e. \\(\pi^2 = \pi = \pi \circ \pi\\)), we need to show that \\(\pi\\) being idempotent implies the same for \\(\text{id}_V\\), and then that the reverse is also true.\
Step 1: establish \\(\rightarrow\\). \
Step 1a: if \\(\text{id}_V - \pi\\) is a projection, then it is idempotent, so: \\(\text{id}_V - \pi = (\text{id}_V - \pi)^2\\). \
Step 1b: expand the RHS: \\(\text{id}_V - \pi = \text{id}_V - 2 \text{id}_V\pi + \pi^2\\). Note that \\(\text{id}_V\\) is itself idempotent, so we simplify by removing the square. \
Step 1c: in assuming it's a projection (since we are proving \\(\rightarrow\\), we assume \\(\pi\\) is idempotent. Then: \\(\text{id}_V - \pi = \text{id}_V^2 - 2\pi + \pi\\). \
Step 1d: Simplify to \\(\text{id}_V = \text{id}_V\\). And we're done. 

Step 2: establish \\(\leftarrow\\). \
Step 2a: if \\(\pi\\) is a projection, then it is idempotent, so need to prove: \\(\pi = \pi^2\\). \
Step 2b: given the same expansion from 1b earlier: \\(\text{id}_V - \pi = \text{id}_V - 2 \text{id}_V\pi + \pi^2\\). \
Step 2c: then (subtracting \\(\text{id}_V\\) and adding \\(2\pi\\) to both sides), we get: \\(\pi = \pi^2\\). And we're done.

b. *Assume now that \\(\pi\\) is a projection. Calculate \\(\text{Im}(\text{id}_V - \pi), \text{ker}(\text{id}_V - \pi)\\) as functions of \\(\text{Im}(\pi), \text{ker}(\pi)\\).*

Approach and key insight: an arbitrary vector \\(v\in V = \pi(v)\in\text{Im}(\pi) + (v - \pi(v))\in\text{ker}(\pi)\\). \
Step 1: calculate \\((\text{id}_V - \pi)(v) = (\text{id}_V - \pi)(\pi(u))\\) for some \\(u\in V\\) s.t. \\(v = \pi(u)\\) (where the identity is just from the definition of \\(v\\).
Step 1a: plug in \\(\pi(u)\\) for \\(v\\): \\((\text{id}_V - \pi)(\pi(u) = (\text{id}_V - \pi)(\pi(u))\\).
Step 1b: simplify by distributing: \\(\pi(u) - \pi(\pi(u)) = \pi(u) - \pi(\pi(u)) = 0\\). 
